src/main/scala/io/github/kiemlicz/shelm/ChartLocation.scala

```diff
@@ -48,44 +49,128 @@ object ChartLocation {
   ) extends ChartLocation
 
   /**
-    * Any remote repository
+    * Any remote legacy repository
     *
-    * @param uri      repo URI
-    * @param settings mainly auth settings
+    * @param uri  repo URI
+    * @param auth mainly auth settings
     */
   case class RemoteRepository(
     chartName: ChartName,
     uri: URI,
-    settings: ChartRepositorySettings,
+    auth: ChartRepositoryAuth,
     chartVersion: Option[String] = None,
   ) extends ChartLocation
+
+  /**
+    *
+    * @param chartName chart name within registry, will be appended as a last path segment
+    * @param uri       registry URL, e.g. oci://registry-1.docker.io/kiemlicz
+    */
+  case class RemoteOciRegistry(
+    chartName: ChartName,
+    uri: URI,
+    chartVersion: Option[String] = None,
+  ) extends ChartLocation
+
 }
 
 case class ChartRepositoryName(name: String) extends AnyVal
 
+object ChartRepositoryName {
+  implicit val decoder: Decoder[ChartRepositoryName] = _.downField("name").as[String].map(ChartRepositoryName(_))
+  implicit val encoder: Encoder[ChartRepositoryName] = crn => Json.obj(("name", Json.fromString(crn.name)))
+}
+
 /**
-  * Helm Chart Repository
+  * `helm repo list -o yaml` single list entry
   */
-case class ChartRepository(
+case class RepoListEntry(chartRepositoryName: ChartRepositoryName, uri: URI)
+
+object RepoListEntry {
+  implicit val decoder: Decoder[RepoListEntry] = c => for {
+    name <- c.as[ChartRepositoryName]
+    uri <- c.downField("url").as[URI]
+  } yield RepoListEntry(name, uri)
+
+  implicit val encoder: Encoder[RepoListEntry] = rle => Json.obj(
+    ("name", Json.fromString(rle.chartRepositoryName.name)),
+    ("url", Json.fromString(rle.uri.toString)),
+  )
+}
+
+/**
+  * Registry/Repository which need prior setup
+  */
+trait ChartRepo {
+  def uri(): URI
+
+  def auth(): ChartRepositoryAuth
+}
+
+trait LegacyRepo extends ChartRepo {
+  def name(): ChartRepositoryName
+}
+
+/**
+  * Ivy repository, .e.g. Artifactory (uses HTTP PUT to upload artifacts)
+  * org.apache.ivy.plugins.resolver.RepositoryResolver#put(org.apache.ivy.core.module.descriptor.Artifact, java.io.File, java.lang.String, boolean)
+  * Helm Chart Repository (!= registry according to Helm doc)
+  *
+  * @param uri repo base url `helm repo add <uri>`
+  */
+case class IvyCompatibleHttpChartRepository(
   name: ChartRepositoryName,
   uri: URI,
-  settings: ChartRepositorySettings = ChartRepositorySettings.NoAuth,
-)
+  auth: ChartRepositoryAuth = ChartRepositoryAuth.NoAuth,
+) extends LegacyRepo
+
+case class ChartMuseumRepository(
```

Comment:
Invalid, see https://github.com/kiemlicz/shelm/pull/16#discussion_r1472526908

Reply:
First things first: absolutely, but...

- `helm push` is OCI only
- `IvyCompatibleHttpChartRepository` and `ChartMuseumRepository` are incompatible, in order to push 'like previously' I've removed the need to specify `IvyCompatibleHttpChartRepository` within `publishToHosting`
-- it's impossible to avoid exposing this distinction to the user, the configuration is far too divergent 
- ultimately the goal is to have OCI only

Check also my last comment, where I've explained rationale